### PR TITLE
runtime: run syscall/js finalizers on wasm without a manual GC

### PR DESCRIPTION
**Pull request by felipegenef**
## Problem

PR #5521 added finalizer support for the block GC. This lets `syscall/js` release bridge table slots when Go objects become unreachable.

TinyGo's block GC normally starts when allocation exhausts the heap. See [`gc_blocks.go`](https://github.com/tinygo-org/tinygo/blob/dev/src/runtime/gc_blocks.go). A `js.Value` uses little Go memory but can keep a JavaScript object and bridge slot alive. The bridge table can therefore grow while the Go heap stays small.

Standard Go also starts a periodic GC from `sysmon`, but its WebAssembly build disables `sysmon`. See [`runtime/proc.go`](https://go.dev/src/runtime/proc.go).

## Changes

- Count net new finalizer registrations after each GC.
- Start pressure GC when the count reaches `max(32, numFinalizers/2)`.
- Run pressure GC at a top level idle point or before a top level WebAssembly export returns.
- Use one bitmap bit per heap block to skip scans for objects without a finalizer.
- Clear finished asyncify stacks and argument bundles after finalizer use.
- Keep blocked task stacks as GC roots.
- Resume queued work after a top level WebAssembly export on JavaScript. Nested exports use the active outer scheduler.

## Scope

The pressure GC applies to the block GC with the `asyncify` and `tasks` schedulers. Stack cleanup applies only to `asyncify`. Other schedulers continue to run finalizers after normal GC cycles.

This change does not modify the public API, finalizer semantics, or `wasm_exec.js`.

## Tests

- Finalizer registration, replacement, removal, and heap address reuse.
- Idle collection without `runtime.GC()`.
- Blocked and completed task stacks.
- Completed task argument bundles.
- Large finalized objects and `runtime_asserts` checks.
- Scheduler build constraint coverage.
- Top level and nested WebAssembly exports.

## Result

Manual WebAssembly stress tests keep the `syscall/js` bridge table bounded without a call to `runtime.GC()`. Binary size checks remain unchanged. Nested exports do not schedule a redundant JavaScript wakeup.

### CODE REVIEW

**Comment by deadprogram** [comment]
Here are some lightly edited generated comments:

# PR #5545 Review Comments

## 1. `finishing` global vs per-task field

The `var finishing bool` in `task_asyncify.go` is a plain package-level global, justified by the cooperative scheduler guarantee that no suspension point exists between `deadlock()`→`Pause()`→`Resume()`. This is correct today, but fragile: if a preemption point is ever introduced in that path, two goroutines could clobber each other's flag silently.

Consider moving this to a field on `*Task` (or a bit in an existing field). That would make the intent explicit ("this specific task is finishing") and eliminate the implicit coupling to cooperative scheduling semantics. The cost is one extra byte or bit per task, which is negligible compared to the stack buffer it already owns.

## 2. Build tag coverage across scheduler variants

The PR splits `gc_finalizer_sched.go` into two files:

- `gc_finalizer_sched.go`: `(gc.conservative || gc.precise) && (scheduler.tasks || scheduler.asyncify)` which installs the idle hook + spawns runner
- `gc_finalizer_sched_other.go`: `(gc.conservative || gc.precise) && !scheduler.none && !scheduler.tasks && !scheduler.asyncify` which spawns runner only

The existing `scheduler.none` path has no `spawnFinalizerRunner` (it drains inline via `wakeFinalizer`).

Please confirm the union of these three paths is exhaustive for every scheduler variant (`none`, `tasks`, `asyncify`, `cores`, `threads`). In particular, verify that no future scheduler variant could fall through all three build constraints and leave `spawnFinalizerRunner` undefined.

## 3. Threshold of 32 regarding tunability and documentation

`finalizerGCThreshold = 32` is described as "a policy constant, like Go's `forcegcperiod`." A few points worth discussing:

- **Is 32 the right default?** For a bursty `syscall/js` workload creating hundreds of `js.Value`s per event, 32 triggers roughly one collection per event handler dispatch (at the idle point), which seems well-calibrated. But a workload that legitimately registers many long-lived finalizers in a setup phase would pay for several unnecessary collections during init.

- **Should it be runtime-tunable?** Go's `GOGC` is an environment variable; `forcegcperiod` is a named constant but not user-facing. If this stays a compile-time constant, documenting it in a comment (already done) is probably sufficient. If there's any chance users will need to tune it, exposing it via an environment variable or `runtime/debug` style API would be worth considering now rather than later.

- **Zero disables the trigger** this is documented in the code comment, which is good. Worth calling out in any user-facing docs if they exist.

**Comment by felipegenef** [comment]
Thanks @deadprogram, all fair points. Done:

**1. `finishing` flag.** Now a per-goroutine field on the asyncify `state` struct instead of a package global, so each finishing goroutine owns its own and the handoff to `Resume` no longer depends on scheduler timing. One byte per task, asyncify only.

**2. Build tag coverage.** There's a fourth file, `gc_finalizer_sched_none.go` (no-op for `scheduler.none`). The three constraints partition the scheduler space, so every variant, present or future, matches exactly one file (`cores`/`threads` land in the `!none && !tasks && !asyncify` catch-all). I documented the partition in a comment and added a test that iterates `validSchedulerOptions`, so a newly added scheduler is checked automatically.

**3. Threshold of 32.** Kept it a `const`. It only fires at an idle point and resets the counter, so a big setup burst costs one collection, not one per 32:

finalizer GC threshold

Happy to make it tunable via a `runtime/debug` setter or an env var if you'd rather; I left the API shape to you since it can be added later without breaking anything.

**Comment by deadprogram** [comment]
Any further comments from @jakebailey @dgryski or anyone else also wasm-involved?

File: src/internal/task/task_asyncify.go [hunk 1]
```diff
@@ -126,6 +162,13 @@ func (t *Task) Resume() {
 	if uintptr(t.state.asyncifysp) > uintptr(t.state.csp) {
 		runtimePanic("stack overflow")
 	}
+	if t.state.finishing {
+		// The goroutine just ran to completion and paused for the last time. It
+		// will never be resumed, so its stack can be cleared now to drop any
+		// pointers its returned frames left behind (see clearStack).
+		t.state.finishing = false
+		t.clearStack()
```
Comment (jakebailey) [inline src/internal/task/task_asyncify.go:152]: I think this might need `t.state.args = nil`?

Comment (felipegenef) [reply src/internal/task/task_asyncify.go:152]: Good catch, fixed. Added a testFinishedGoroutineArgs case in finalizeridle.go to cover it. Thanks!

**Comment by felipegenef** [comment]
@deadprogram I kept thinking about your question on the 32 threshold, so I went and read how upstream Go paces this. It ties collection to proportional growth (GOGC) and keeps a per-span bitmap so the collector can skip spans with nothing registered, rather than relying on a fixed count.

Borrowing both on top of what's already here:

- `max(32, numFinalizers/2)` as the trigger, which is the GOGC reasoning and the direct answer to your question: the cost of any fixed threshold depends on how much is registered, so a proportion travels better than a constant. 32 becomes the floor.
- a per-block registration bit, the analogue of Go's span bitmap, so a fresh object answers "already registered?" with a bit test. Lazily allocated, grown with the heap, and conservative for addresses it doesn't cover yet.

Both mechanisms are battle-tested in Go's own collector, which counts for something this central. I benchmarked the branch before and after on a WASM workload that registers finalizers heavily: the new one is consistently faster and, more importantly, stops degrading as the table grows. Reclamation behaviour is unchanged.

I also tried Go's per-span partitioning (64-way buckets) and measured it as redundant here: with the bit there's no scan left to shorten, and the partitioning mostly buys per-span locking, which doesn't apply under a global lock. Left it out.

`testdata/finalizerbits.go` covers clear-then-register, register-twice-replaces, register/clear churn, allocation into reused memory and a mixed batch. They pass with and without the change, by design.

Merged current dev, and `TestBinarySize` comes out unchanged on all three targets, so nothing to adjust there.

**Comment by felipegenef** [comment]
Hey, small ping on this one ☺️.

I've been using this branch as the WASM backend engine for a Go web framework I maintain, and it's been holding up really nicely so far. Even on the rougher stress tests (10k clicks on a page over 5 minutes) the slot count stays flat, which was exactly the thing I was hoping to fix.

No rush at all, just didn't want it to slip off the radar. Thanks for the reviews so far!

@deadprogram @jakebailey

**Comment by deadprogram** [comment]
@jakebailey any more comments on this PR?

File: src/runtime/gc_finalizer.go [hunk 2]
```diff
@@ -66,9 +181,18 @@ func decodeFinalizerPtr(enc uintptr) uintptr  { return ^enc }
 func registerFinalizer(addr uintptr, fn interface{}) {
 	enc := encodeFinalizerPtr(addr)
 
+	tracked := isOnHeap(addr)
+
 	if fn == nil {
-		// Clear: remove every registration for this object.
+		// Clear: remove every registration for this object. The bit proves in
+		// one test that there is nothing to remove.
+		if tracked && !finalizerBitGet(addr) {
```
Comment (jakebailey) [inline src/runtime/gc_finalizer.go:171]: The clear fast path reads `finalizerBits` without holding `gcLock`, while registration can concurrently resize the slice or update its bytes. This code is also built for the parallel `cores`/`threads` schedulers, so `SetFinalizer(obj, nil)` could incorrectly observe an unset bit and leave the finalizer registered.

Could the bitmap check be moved under `gcLock`, keeping the check, bit clear, and table removal in the same critical section?

**Comment by dgryski** [comment]
If this code works on other platforms/schedulers than just javascript, it would be nice to have some tests demonstrating that.

File: src/runtime/gc_finalizer.go (hunk 2 above)
Comment (felipegenef) [reply src/runtime/gc_finalizer.go:171]: Fixed now.

Moved the check, the bit clear and the table removal into the same critical section. `isOnHeap` went in with them.

I found two more while I was there: the slice header swap in `adoptFinalizerBits`, and the sizing read in `growFinalizerBits` (that one's now `finalizerBitsShortfall` and runs under the lock).

This shouldn't cost anything. The bit was there to skip the O(numFinalizers) walk, not the lock, and the clear path was taking `gcLock` two lines down anyway.

Rebuilt on wasm, wasip1, microbit, pico, pico+cores and threads. wasm finalizer tests still pass.

**Comment by felipegenef** [comment]
@dgryski I added some tests for the other platforms.

`testdata/finalizerinvariants.go` runs on every scheduler now: asyncify (wasm), tasks (cortex-m-qemu), cores (riscv-qemu), plus threads and none on the host with `-gc=conservative`.

I kept it to assertions that hold under conservative stack scanning: a cleared finalizer never runs, a replaced one never runs the old func, a reachable object is never finalized, nothing runs twice. Those are the ones the registration bitmap can break.

What it doesn't assert is that a finalizer actually fired. That needs the dropped object to really get collected, which only holds on wasm, so the three existing tests stay wasm-only.

AVR I had to skip. Registering and clearing works fine there, but a single `runtime.GC()` never comes back, and `gc.go` is already skipped on simavr for the same reason.

As for whether other platforms actually gain anything, honestly I think it's just WASM for now. `syscall/js` is the only place in the tree that registers a finalizer at all. That might change though. TinyGo has its own `os` package, and unlike upstream Go its `os.File` doesn't close the fd on GC. If that ever lands, WASI would pick this up for free, since it's already on the block GC with a cooperative idle point.

**Comment by felipegenef** [comment]
Had to make a small change after the Windows CI broke: the two host scheduler variants I added don't build there. `scheduler.threads` needs `threadID`, which `internal/task` only defines for linux and darwin, and `scheduler.none` doesn't link either. Both reproduce with any testdata file, so they aren't new.

Limited those two runs to linux and darwin, same as `TestTimerStopResetRace` already does.

File: testdata/finalizerinvariants.go [hunk 3]
```diff
@@ -0,0 +1,139 @@
...
+
+	for i := 0; i < batch; i++ {
+		keepReachable(i)
+	}
+
+	// Collect repeatedly, yielding so the finalizer runner goroutine gets to
+	// drain anything that was queued.
+	for i := 0; i < 4; i++ {
+		runtime.GC()
+		runtime.Gosched()
```
Comment (jakebailey) [inline testdata/finalizerinvariants.go]: `runtime.Gosched()` is a no-op under `scheduler.threads`, so these loops do not ensure the finalizer runner has drained before checking the counters.

Comment (felipegenef) [reply testdata/finalizerinvariants.go]: Got it.

`drainFinalizers` now polls until no new finalizer runs for three rounds in a row, instead of yielding a fixed number of times. Every target drains in 4 rounds and then 3.

I also made the test fail if no finalizer ran at all, so it can't pass on an empty state. That caught the plain host runs on boehm, which are skipped now.

**Comment by felipegenef** [comment]
All Tests passed. 

Thanks for the reviews so far @deadprogram @jakebailey @dgryski .

Let me know if you have anything else.

File: testdata/finalizerinvariants.go [hunk 4]
```diff
@@ -0,0 +1,209 @@
...
+	p := &obj{x: id}
+	runtime.SetFinalizer(p, func(*obj) { reachedRan++ })
+	reachable = append(reachable, p)
+}
+
+// finalizerRuns is the total number of finalizer invocations observed so far,
+// across every counter. Individual counters are asserted on at the end; this
+// sum exists only to tell "the runner is still working" from "the queue is
+// empty".
+func finalizerRuns() int {
```
Comment (jakebailey) [inline testdata/finalizerinvariants.go]: These counters are still read concurrently with finalizer-thread writes.

File: testdata/finalizerinvariants.go [hunk 5]
```diff
@@ -0,0 +1,209 @@
...
+		time.Sleep(time.Millisecond)
+		switch {
+		case finalizerRuns() != before:
+			quiet = 0
+		case before == 0:
+			// Nothing has run yet: cannot tell a drained queue from a runner
+			// that has not started, so keep polling.
+		default:
+			quiet++
+			if quiet == quietRounds {
```
Comment (jakebailey) [inline testdata/finalizerinvariants.go]: I don’t think polling like this necessarily fixes the problem; the runner could simply be delayed during the quiet rounds.

File: testdata/finalizerinvariants.go (hunk 4 above)
Comment (felipegenef) [reply testdata/finalizerinvariants.go]: Made them `sync/atomic`. Should be solved now.

File: testdata/finalizerinvariants.go (hunk 5 above)
Comment (felipegenef) [reply testdata/finalizerinvariants.go]: Fair. It counts now instead of waiting for things to go quiet, so the loop keeps collecting until all 8 replacement finalizers have run, and fails if it never gets there.

**Comment by dgryski** [comment]
The gc and scheduler, if build with `-tags=runtime_asserts`, set `gcAsserts = true` and `schedulerAsserts = true`, which are used to validate data structures and other things to ensure there are no invalid states with their bookkeeping or usage.  Are there any asserts we could add to the new finalizer code?

**Comment by felipegenef** [comment]
@dgryski  Added some asserts for the finalizer bookkeeping.

The main ones check the registration bitmap against the table it stands in for: when the bit says "not registered" and the clear or replace path takes the shortcut, the assert walks the table anyway and fails if the object is actually there.

The others cover `numFinalizers` matching the table, every pending object surviving resurrection, and nothing being dequeued after it's been freed.

**Comment by felipegenef** [comment]
Hey all, just a friendly check-in on this one.

All review comments are addressed and CI is green. The new cross-scheduler tests cover asyncify, tasks, cores and threads, and the locking fix from the last round is verified on wasm, wasip1, microbit and pico. The runtime_asserts coverage for the finalizer bookkeeping landed as well.

This branch has been serving as the wasm backend for a framework I maintain, and the bridge table stays flat under sustained load, which was the whole point of the PR.

A fresh pass would be really welcome when you have time. Thanks for all the careful reviews so far.

@deadprogram @jakebailey @dgryski

**Comment by deadprogram** [comment]
@felipegenef here are some edited feedback items from a further automated review based on your latest changes:

1. The stack zeroing isn't gated on finalizer use. `MarkFinishing()` is called unconditionally from `deadlock()` and `goexit()`, and asyncify's Resume unconditionally memzeros the whole stack buffer.

 2. `deadlock()` catches more than finished goroutines. Nil-channel chanSend/chanRecv (chan.go:234, chan.go:314) and select {} (compiler/channel.go:92) all route through it, so a stuck goroutine gets its frame wiped too.

  3. finalizerinvariants.go hard-fails on a positive-collection precondition. The header explains the file exists precisely because conservative stack scanning elsewhere can legitimately keep a dropped object alive but then waitForDrain() requires all 8 dropReplaced finalizers to fire and reports FAIL otherwise.

  4. The new test-matrix loop is broader than the comment says. It's gated on options.Target == "", but optionsFromOSARCH also leaves Target empty, so it fires for linux/386, linux/arm/6, linux/arm64 and linux/mips

  5. Stale comment on finalizerGCThreshold. It still says the trigger "caps that accumulation at roughly this many entries regardless of heap size"

  6. finalizersSinceGC never decrements on clear.

**Comment by felipegenef** [comment]
@deadprogram Thanks, all six items are addressed:

1. Stack zeroing is now gated on finalizer use: `MarkFinishing()` is called from normal completion and `goexit()` only after `finalizerIdleGC` has been installed.
2. Normal asyncify completion now uses `runtime.exitGoroutine`, while true deadlocks keep `select {}` and nil-channel send/receive tasks rooted without clearing their stacks. The regression test covers all three paths and performs an additional GC/drain after proving finalizer progress.
3. `finalizerinvariants.go` no longer requires positive collection on conservative targets. It checks only forbidden outcomes and enables the existing `runtime_asserts` bookkeeping checks.
4. The `threads`/`none` variants now run only inside the explicit host subtest, so `optionsFromOSARCH` cross-targets no longer trigger them.
5. The threshold comments now describe `finalizerGCThreshold` as the floor raised proportionally by `finalizerGCTrigger`.
6. Clearing an actual registration now decrements `finalizersSinceGC`, saturating at zero, so clears offset net registration pressure.

Validation completed: the full test suite passes, the blocked-stack test passes 100 consecutive runs, the host invariant variants pass 50 consecutive runs, compiler LLVM goldens pass, and `TestBinarySize` is unchanged.

**Comment by felipegenef** [comment]
@deadprogram Thanks, all six items are addressed:

1. Stack zeroing is gated on finalizer use for normal asyncify completion, `goexit()`, and scheduler-enabled `//go:wasmexport` completion.
2. Normal completion uses `runtime.exitGoroutine`, while `select {}` and nil-channel send/receive deadlocks remain rooted without clearing their stacks.
3. `finalizerinvariants.go` checks only portable negative invariants and no longer requires positive collection on conservative targets.
4. The `threads`/`none` variants run only in the explicit host subtest.
5. The threshold comments now describe the proportional trigger correctly.
6. Clearing a registration decrements `finalizersSinceGC`, saturating at zero.

I also added focused coverage for wasm-export completion and multi-block finalized objects. CI is green, including compiler and binary-size checks.

**Comment by deadprogram** [comment]
1. Unrelated fix hidden behind the finalizer gate `scheduler_cooperative.go:296-306`

```
if task.Current() == nil && finalizerIdleGC != nil {
    finalizerIdleGC()
    if asyncScheduler && (!runqueue.Empty() || sleepQueue != nil || timerQueue != nil) {
        sleepTicks(0)
    }
}
```

The sleepTicks(0) re-arm is what lets goroutines started inside a //go:wasmexport resume after the export returns. That has nothing to do with finalizers, but it only runs if a finalizer was registered. Pull it out of the finalizerIdleGC != nil check, or state in the comment that it's deliberately scoped for now.

Same thing is in exitGoroutine/goexit/wasmExportExit: if finalizerIdleGC != nil { task.MarkFinishing() }. Add a comment of explantion.

2. Dropped rationale in `main_test.go`

The host-scheduler block moved but lost the comment explaining the switch runtime.GOOS: threads needs threadID, which internal/task defines only on linux/darwin, and none doesn't link on Windows. Add comment so the next person hitting a Windows failure doesn't have to rediscover it.

**Comment by felipegenef** [comment]
@deadprogram both items are addressed.

- Moved `sleepTicks(0)` outside the finalizer gate and added coverage for scheduler resumption before any finalizer is registered.
- Added the requested comments explaining the `MarkFinishing` gates and the Linux/Darwin test restriction.

The focused WASM test passes.

**Comment by deadprogram** [comment]
One new note:

sleepTicks(0) is now also reached when the export ran re-entrantly, i.e. task.Current() != nil after t.Resume() since Resume saves and restores currentTask (task_asyncify.go:143,157), This nesting happens whenever JS calls into an export from inside a js.Func callback. There, the outer scheduler will drain the run queue on its own, so the extra sleepTicks(0) just queues a redundant setTimeout → go_scheduler() (wasm_exec.js:272-282, no dedup).

This results in one spare macrotask per nested export call, which under callback-heavy syscall/js traffic is not nothing. Could avoid by keeping the re-arm under the same task.Current() == nil check as the GC.

**Comment by felipegenef** [comment]
@deadprogram Thanks, fixed.

The re-arm now remains under `task.Current() == nil`, while staying independent of `finalizerIdleGC`.

I also added coverage using a real `js.Func` → JavaScript → nested `//go:wasmexport` path. The test verifies one wakeup for a top-level export and none for a re-entrant export. The focused WASM regression passes.

The `smoke-test-windows` job failed during `Set up job` because GitHub could not download `setup-scoop` due to an internal DNS error, before checkout or execution of repository code. I do not have permission to re-run the workflow.

**Comment by deadprogram** [comment]
@felipegenef one other thing, perhaps you can reduce the size/amount of comments in this pr a bit?

I suggest try to follow these guidelines:

## Comments
Use ASD-STE100 Simplified Technical English for content outside of the code like comments, PR description, PR title, github comments, and issues.

Generally omit extra redundant comments. If a comment is absolutely needed, always keep it brief, no more than 2 lines. Avoid extra use of colons, semicolons, and dashes. When comments are required due to some specific requirements, include a reference such as to a datasheet or other definitive source to explain why this is the case, including the chapter/section number, page number, and/or URL.

**Comment by felipegenef** [comment]
@deadprogram Sure. Sorry for the long comments. This change is important to TinyGo's WASM support and essential to a project I maintain, where TinyGo is the main WASM engine.

Because this PR matters a lot to me, I wanted to explain my reasoning clearly. English is not my first language, and some GC terms are outside my current english technical vocabulary, so I used AI to help with the wording. The result was more verbose than needed, and I should have reviewed it more carefully against the project style.

Thanks for the feedback. I removed comments that were not needed and shortened the remaining comments.

**Comment by deadprogram** [comment]
Thank you very much for all your work on this @felipegenef and to everyone who helped review. Now squash/merging!
